README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

## Simple Tabulation Made Simple

**Status**

*lines of R code:* 150, *lines of test code:* 117

[![repo
status](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
<a href="https://travis-ci.org/petermeissner/tabit">
<img src="https://api.travis-ci.org/petermeissner/tabit.svg?branch=master">
<a/> [![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/petermeissner/tabit?branch=master&svg=true)](https://ci.appveyor.com/project/petermeissner/ical)
<a href="https://cran.r-project.org/package=tabit">
<img src="http://www.r-pkg.org/badges/version/tabit"> </a>
<a href="https://codecov.io/gh/petermeissner/tabit">
<img src="https://codecov.io/gh/petermeissner/tabit/branch/master/graph/badge.svg" alt="Codecov" />
</a> <img src="http://cranlogs.r-pkg.org/badges/grand-total/tabit">
<img src="http://cranlogs.r-pkg.org/badges/tabit">

**Development version**

0.2.1 - 2020-03-30 / 20:12:13

**Description**

Simple tabulation should be dead simple. This package is an opinionated
approach to easy tabulations while also providing exact numbers and
allowing for re-usability. This is achieved by providing tabulations as
data.frames with columns for values, optional variable names, frequency
counts including and excluding NAs and percentages for counts including
and excluding NAs. Also values are automatically sorted by in decreasing
order of frequency counts to allow for fast skimming of the most
important information.

**License**

MIT + file LICENSE <br>Peter Meissner \[aut, cre\]

## Installation

You can install the released version of tabit from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("tabit")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("petermeissner/tabit")
```

## Example

``` r
# load package
library(tabit)
```

**simple vectors**

``` r
# simple tabulation
tabit(mtcars$cyl)
##   .variable .value .count  .pct .pct_incl_na
## 3         .      8     14 43.75        43.75
## 1         .      4     11 34.38        34.38
## 2         .      6      7 21.88        21.88
## 4         .     NA      0    NA         0.00
```

**simple data.frames**

``` r
# can we do this with data.frames
tabit(mtcars[, c("cyl", "am")])
##   cyl am .count  .pct .pct_incl_na
## 3   8  0     12 37.50        37.50
## 4   4  1      8 25.00        25.00
## 2   6  0      4 12.50        12.50
## 1   4  0      3  9.38         9.38
## 5   6  1      3  9.38         9.38
## 6   8  1      2  6.25         6.25
```

**simple tibbles**

``` r
# what about this tibble thing?
suppressPackageStartupMessages({
  library(dplyr)
})
## Warning: package 'dplyr' was built under R version 3.6.3
mtcars %>% 
  select(cyl, am) %>% 
  tabit()
##   cyl am .count  .pct .pct_incl_na
## 3   8  0     12 37.50        37.50
## 4   4  1      8 25.00        25.00
## 2   6  0      4 12.50        12.50
## 1   4  0      3  9.38         9.38
## 5   6  1      3  9.38         9.38
## 6   8  1      2  6.25         6.25
```

**grouped tibbles**

``` r
# ... and grouped tibbles?
mtcars %>% 
  group_by(cyl, am) %>% 
  tabit()
##   cyl am .count  .pct .pct_incl_na
## 3   8  0     12 37.50        37.50
## 4   4  1      8 25.00        25.00
## 2   6  0      4 12.50        12.50
## 1   4  0      3  9.38         9.38
## 5   6  1      3  9.38         9.38
## 6   8  1      2  6.25         6.25
```

**too much magic?**

``` r
# use tabit_1() for vector usage
tabit_1(letters[1:4])
##   .variable .value .count .pct .pct_incl_na
## 1         .      a      1   25           25
## 2         .      b      1   25           25
## 3         .      c      1   25           25
## 4         .      d      1   25           25
## 5         .     NA      0   NA            0

# use tabit_x for data.frames and tibbles
tabit_x(mtcars[, c("cyl", "am")])
##   cyl am .count  .pct .pct_incl_na
## 3   8  0     12 37.50        37.50
## 4   4  1      8 25.00        25.00
## 2   6  0      4 12.50        12.50
## 1   4  0      3  9.38         9.38
## 5   6  1      3  9.38         9.38
## 6   8  1      2  6.25         6.25
```


NEWS.md
# tabit 0.2.1 - 2020-03-30

* adding tabit_x()
* moving tabit() code to tabit_1()
* making tabit() a magic function doing tabit_1() on vectors and lists and 
  tabit_x() on data.frames

# tabit 0.1.0 & 0.1.1

* Initial CRAN release
* Initial version
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(tabit)

test_check("tabit")


tests/testthat/test-tabit_1.R
test_that(
  "tabit_1",
  {
    expect_true(
      nrow(tabit_1(1)) == 2
    )

    expect_true(
      nrow(
        tabit_1(1, useNA = "no")
      ) == 1
    )

    expect_true(
      nrow(
        tabit_1(1, useNA = NULL)
      ) == 1
    )

    expect_true(
      nrow(
        tabit_1(NA, useNA = NULL)
      ) == 0
    )

    expect_true(
      nrow(
        tabit_1(list(), useNA = NULL)
      ) == 0
    )

    expect_true(
      ncol(
        tabit_1(list(), useNA = NULL)
      ) == 5
    )

    expect_true(
      is.null(
        ncol(
          tabit_1(list(), useNA = NULL, as_df  = FALSE)
        )
      )
    )

    expect_true(
      ncol(
        tabit_1(data.frame(a = 1:10, b = 0))
      ) == 5
    )
  }
)


tests/testthat/test-tabit_x.R
test_that(
  "tabit_x",
  {
    expect_error(
      tabit_x(1)
    )

    expect_error(
      tabit_x(matrix(1))
    )

    expect_true(
      nrow(
        tabit_x(data.frame(a=1))
      ) == 1
    )

    expect_true(
      nrow(
        tabit_x(data.frame(a=c(1, NA)))
      ) == 2
    )

    expect_true(
      nrow(
        tabit_x(data.frame(a = character()))
      ) == 0
    )

    expect_true(
      nrow(
        tabit_x(data.frame())
      ) == 0
    )

    expect_true(
      ncol(
        tabit_x(data.frame())
      ) == 1
    )

    expect_true(
      ncol(
        tabit_x(data.frame(a = integer(0)))
      ) == 2
    )


    expect_true(
        tabit_x(
          x = data.frame(a = c(1,2,2,3))
        )$.count[1] == 2
    )

    expect_true(
      tabit_x(
        x = data.frame(a = c(1,2,2,3)),
        sort = 0
      )$a[1] == 1
    )

    expect_true(
      tabit_x(
        x = data.frame(a = c(1,2,2,3)),
        sort = -1
      )$a[3] == 2
    )

    expect_true(
      tabit_x(
        x = data.frame(a = c(1,2,2,3)),
        sort = -1
      )$.count[3] == 2
    )

    expect_true(
      tabit_x(
        x = data.frame(a = c(1,2,2,3))
      )$a[1] == 2
    )

  }
)


R/tab1.R
#' ti_tab1
#'
#' @param x the thing to be tabulated
#' @param sort should
#' @param digits the number of digits to round percentages to
#' @param ... further optional arguments passed through to methods
#' @param as_df if more than one vector has been tabulated at once, should
#'     the results be combined into a data.frame before returning
#'
#' @export
#'
#' @return Returns a data frame with columns
#'  \code{value} (the values tabulated by),
#'  \code{count} (the number of times a value occured),
#'  \code{pct} (the percentage that value occurred excluding NAs) and
#'  \code{pct_all} (the percentage that value occurred including NAs)
#'
#'
#'
#' @examples
#'
#' ti_tab1(mtcars$cyl)
#' ti_tab1(mtcars[, c("cyl", "am")])
#'
#'
ti_tab1 <-
  function (x, sort = 1, digits = 2, ...) {
    UseMethod("ti_tab1", x)
  }



#' ti_tab1.data.frame
#'
#' @rdname ti_tab1
#'
#' @export
#'
ti_tab1.data.frame <-
  function (x, sort = 1, digits = 2, ..., as_df = TRUE ) {
    ti_tab1(as.list(x), sort = sort, digits = digits, ..., as_df = as_df)
  }



#' ti_tab1.list
#'
#' @rdname ti_tab1
#'
#' @export
#'
ti_tab1.list <-
  function (x, sort = 1, digits = 2, ..., as_df = TRUE) {
    tmp <- lapply(x, ti_tab1, sort = sort, ...)

    if ( as_df == TRUE ){
      nam <- names(x)
      for ( i in seq_along(nam) ){
        tmp[[i]] <-
          cbind(
            data.frame(name = nam[i]),
            tmp[[i]]
          )
      }
      tmp            <- do.call(rbind, tmp)
      row.names(tmp) <- NULL
      return(tmp)
    } else {
      return(tmp)
    }
  }



#' ti_tab1.default
#'
#' @rdname ti_tab1
#'
#' @export
#'
ti_tab1.default <-
  function (x, sort = 1, digits = 2, ...) {
    tmp     <- table(x, useNA = "always")
    tmp_val <- as.integer(tmp)
    tmp_nam <- names(tmp)

    tmp <-
      data.frame(
        value   = tmp_nam,
        count   = tmp_val,
        pct     =
          round(
            c(
              tmp_val[-length(tmp_val)] / sum(tmp_val[-length(tmp_val)]),
              NA
            ) * 100,
            digits
          ),
        pct_all = round(tmp_val / sum(tmp_val) * 100, digits)
      )

    # return - ordered by count in descending
    if ( sort < 0 ){
      return(tmp[order( tmp$count), ])
    } else if ( sort > 0 ){
      return(tmp[order(-tmp$count), ])
    } else {
      return(tmp)
    }
  }


R/tabit_x.R
#' tabit_x
#'
#' Function for cross tabulation of variables within a data.frame or data.frame
#' like.
#'
#' The data.frame method will take all column value combination for cross
#' tabulation while the grouped_df method is explicitly for tibbles with group
#' information. For those only value combinations for the group columns are
#' relevant.
#'
#'
#' @param x the thing to be tabulated
#' @param sort should
#' @param digits the number of digits to round percentages to
#' @param useNA whether ro us NAs as value category or not, defaults to TRUE
#' @param ... further optional arguments passed through to methods
#'
#' @export
#'
#' @return Returns a data frame with columns
#'  zero, one or more columns with value combinations to be cross tabulated by,
#'  \code{.count} (the number of times a value occurred),
#'  \code{.pct} (the percentage that value occurred excluding NAs) and
#'  \code{.pct_incl_na} (the percentage that value occurred including NAs)
#'
#'
#' @import stats
#'
#' @examples
#'
#' tabit_x(mtcars)
#' tabit_x(mtcars[, c("cyl", "am")])
#'
#'
tabit_x <-
  function (x, ..., sort = 1, digits = 2, useNA = TRUE ) {
    UseMethod("tabit_x", x)
  }



#' tabit_x.data.frame
#'
#' @rdname tabit_x
#'
#' @export
#'
tabit_x.data.frame <-
  function (x, ..., sort = 1, digits = 2, useNA = TRUE ) {

    # process by parameter
    by <- x

    # short circuit edge cases
    if ( ncol(by) == 0 ){
      return(
        data.frame(.count = integer(0))
      )
    }

    if ( nrow(by) == 0 ){
      return(
        cbind(
          df    =  by[0,, drop = FALSE],
          .count = integer(0)
        )
      )
    }


    # process useNA parameter
    if ( useNA == TRUE ){
      by <-
        lapply(
          X   = by,
          FUN =
            function(by){
              by[is.na(by)] <- "$//NA//$"
              as.character(by)
            }
        )
    }

    # execute aggregation
    tmp <-
      stats::aggregate(rep(1, nrow(x)), FUN = sum, by = by)

    # post process aggregation
    names(tmp)[length(names(tmp))] <- ".count"
    tmp[tmp == "$//NA//$" ] <- NA


    # calculate percentages
    iffer <- complete.cases(tmp)
    tmp[iffer, ".pct"] <-
      round(
        x      = tmp[iffer, ".count"] / sum(tmp[iffer, ".count"]) * 100,
        digits = digits
      )

    tmp$.pct_incl_na <-
      round(
        x      = tmp[, ".count"] / sum(tmp[, ".count"]) * 100,
        digits = digits
      )



    # sort
    if ( sort > 0 ){
      tmp <- tmp[order(-tmp$.count), ]
    } else if ( sort < 0 ){
      tmp <- tmp[order(tmp$.count), ]
    }

    # return
    tmp
  }


#' tabit_x.grouped_df
#'
#' @rdname tabit_x
#'
#' @export
#'
tabit_x.grouped_df <-
  function (x, ..., sort = 1, digits = 2, useNA = TRUE ) {

    # process by parameter
    by_nam <- names(attributes(x)$groups)
    by <- x[by_nam[by_nam != ".rows"]]

    tabit_x.data.frame(by, ..., sort = sort, digits = digits, useNA = useNA)
  }


#' tabit_x.default
#'
#' @rdname tabit_x
#'
#' @export
#'
tabit_x.default <-
  function (x, ..., sort = 1, digits = 2, useNA = "always") {
    stop(
      "tabit_x is not implemented for type: ",
      paste(class(x), collapse = ", ")
    )
  }



R/tabit_1.R
#' tabit_1
#'
#' Function to do frequency tables for single vectors or to do multiple frequency
#' tabulations one for each column.
#'
#' @param x the thing to be tabulated
#' @param sort should
#' @param digits the number of digits to round percentages to
#' @param ... further optional arguments passed through to methods
#' @param as_df if more than one vector has been tabulated at once, should
#'     the results be combined into a data.frame before returning
#' @param na_to_char transform NA values to "NA" character string or not, defaults to TRUE
#' @param pct add percentage columns
#' @param useNA how to handle NAs, defaults to "always" to always include NA as category
#'
#' @export
#'
#' @return Returns a data frame with columns
#'  \code{.variable} (the variable the values stem from),
#'  \code{.value} (the values tabulated by),
#'  \code{.count} (the number of times a value occurred),
#'  \code{.pct} (the percentage that value occurred excluding NAs) and
#'  \code{.pct_incl_na} (the percentage that value occurred including NAs)
#'
#'
#'
#' @examples
#'
#' tabit_1(mtcars$cyl)
#' ##   .variable .value .count     .pct  pct_incl_na
#' ## 3         .      8     14    43.75        43.75
#' ## 1         .      4     11    34.38        34.38
#' ## 2         .      6      7    21.88        21.88
#' ## 4         .     NA      0       NA         0.00
#'
#'
#'
#' tabit_1(mtcars[, c("cyl", "am")])
#' ##   .variable .value .count   .pct  .pct_incl_na
#' ## 1       cyl      8     14  43.75         43.75
#' ## 2       cyl      4     11  34.38         34.38
#' ## 3       cyl      6      7  21.88         21.88
#' ## 4       cyl     NA      0     NA          0.00
#' ## 5        am      0     19  59.38         59.38
#' ## 6        am      1     13  40.62         40.62
#' ## 7        am     NA      0     NA          0.00
#'
#'
tabit_1 <-
  function (x, sort = 1, digits = 2, ...) {
    UseMethod("tabit_1", x)
  }



#' tabit_1.data.frame
#'
#' @rdname tabit_1
#'
#' @export
#'
tabit_1.data.frame <-
  function (x, sort = 1, digits = 2, ..., as_df = TRUE ) {
    tabit_1(as.list(x), sort = sort, digits = digits, ..., as_df = as_df)
  }



#' tabit_1.list
#'
#' @rdname tabit_1
#'
#' @export
#'
tabit_1.list <-
  function (x, sort = 1, digits = 2, ..., as_df = TRUE) {
    x   <- as.data.frame(x,  stringsAsFactors = FALSE)
    tmp <- lapply(x, tabit_1, sort = sort, ...)

    nam <- names(x)
    for ( i in seq_along(nam) ){
      tmp[[i]]$.variable <- nam[i]
    }

    if ( as_df == TRUE ){
      tmp            <-
        do.call(
          what = rbind,
          args =
            c(
              list(
                data.frame(
                  .variable     = character(0),
                  .value        = character(0),
                  .count        = integer(0),
                  .pct          = numeric(0),
                  .pct_incl_na  = numeric(0),
                  stringsAsFactors = FALSE
                )
              ),
              tmp
            )
          )
      row.names(tmp) <- NULL
      return(tmp)
    } else {
      return(tmp)
    }
  }



#' tabit_1.default
#'
#' @rdname tabit_1
#'
#' @export
#'
tabit_1.default <-
  function (x, sort = 1, digits = 2, useNA = "always", pct = TRUE, na_to_char = TRUE, ...) {

    # execute tabulations
    tmp     <- table(x, useNA = useNA)
    tmp_nam <- names(tmp)
    tmp_val <- as.integer(tmp)

    # generate data.frame
    tmp <-
      data.frame(
        .variable = rep(".", length(tmp_val)),
        .value    = tmp_nam,
        .count    = tmp_val,
        stringsAsFactors = FALSE
      )

    # add percentages
    if ( pct == TRUE ){
      sub_set <- !is.na(tmp_nam)
      tmp$.pct[sub_set] <-
        round(tmp_val[sub_set] / sum(tmp_val[sub_set]) * 100, digits)

      tmp$.pct_incl_na <-
        round(tmp_val / sum(tmp_val) * 100, digits)
    }

    # transform NA values to "NA" string
    if ( na_to_char == TRUE ){
      tmp$.value[is.na(tmp$.value)] <- "NA"
    }


    # return - ordered by count, in descending order, or not
    if ( sort < 0 ){
      return(tmp[order( tmp$.count), ])
    } else if ( sort > 0 ){
      return(tmp[order(-tmp$.count), ])
    } else {
      return(tmp)
    }
  }

